### PR TITLE
Moondream2 VisionEncoder

### PR DESCRIPTION
**Pull request by timgestson**
Vision Encoder for the Moondream2 model

### CODE REVIEW

**Comment by timgestson** [comment]
Not exactly sure how to reproduce the error [here](https://github.com/huggingface/ratchet/actions/runs/9235604740/job/25411232525) 
Im running M3 Mac running:
`cargo nextest run -j 1 --no-fail-fast --features=ci,pyo3` 
and the test passes with no issues

**Comment by FL33TW00D** [comment]
> Not exactly sure how to reproduce the error [here](https://github.com/huggingface/ratchet/actions/runs/9235604740/job/25411232525) Im running M3 Mac running: `cargo nextest run -j 1 --no-fail-fast --features=ci,pyo3` and the test passes with no issues

Don't worry about it, it's something to do with the metal virtualization that actions does. I'll fix it!

**Comment by timgestson** [comment]
Quantized and tested in the browser as well.  Moving to the latest checkpoint is a slightly larger effort since there is a lot more [image preprocessing](https://github.com/vikhyat/moondream/blob/main/moondream/vision_encoder.py#L189-L325) to support diff image sizes.  But can be a follow up.

Will we squash on merge or should I rebase all of these commits into 1?

File: crates/ratchet-models/src/registry.rs [hunk 1]
```diff
@@ -75,6 +76,7 @@ impl AvailableModels {
                 PhiVariants::Phi2 => "FL33TW00D-HF/phi2",
                 PhiVariants::Phi3 => "FL33TW00D-HF/phi3",
             },
+            AvailableModels::Moondream => "tgestson/ratchet-moondream2",
```
Comment (timgestson) [inline crates/ratchet-models/src/registry.rs:79]: Do we want to move the weights to a repo owned by you (or the project) to be more uniform?  They are now just created with candles tensor-tools from [this moondream2 model.safetensors file](https://huggingface.co/vikhyatk/moondream2/blob/2024-05-08/model.safetensors)

Comment (FL33TW00D) [reply crates/ratchet-models/src/registry.rs:79]: @timgestson let's create a repo for the project!

File: crates/ratchet-models/src/moondream/model.rs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    )
+    return model.encode_image(tensor).numpy()
+"#;
+
+        run_py_prg(prg.to_string(), &[&tensor], &[], ratchet::DType::F32)
+    }
+
+    #[test]
+    fn vision_encoder() {
+        let device = Device::request_device(DeviceRequest::GPU).unwrap();
```
Comment (FL33TW00D) [inline crates/ratchet-models/src/moondream/model.rs:235]: Can you use the `GPU_DEVICE` like in: https://github.com/huggingface/ratchet/blob/master/crates/ratchet-core/src/ops/softmax.rs#L296

Might help fix the test!

File: crates/ratchet-models/src/moondream/model.rs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+        let api = Api::new().unwrap();
+        let model_repo = api.model("tgestson/ratchet-moondream2".to_string());
+        let model_path = model_repo.get("moondream_f32.gguf").unwrap();
```
Comment (FL33TW00D) [inline crates/ratchet-models/src/moondream/model.rs:239]: Changing this to be smaller might also help fix the test, the metal runners on CI are weak.

File: crates/ratchet-models/src/registry.rs (hunk 1 above)
Comment (timgestson) [reply crates/ratchet-models/src/registry.rs]: Moved to the community repo

File: crates/ratchet-models/src/moondream/model.rs (hunk 2 above)
Comment (timgestson) [reply crates/ratchet-models/src/moondream/model.rs]: Done

File: crates/ratchet-models/src/moondream/model.rs (hunk 3 above)
Comment (timgestson) [reply crates/ratchet-models/src/moondream/model.rs:241]: Unfortunately we won't get a close match with the PyTorch ground truth that operates on f16 if we use the quantized weights.  These gguf weights are already stored in f16 despite the naming.  Maybe for now we use
`#[cfg_attr(feature = "ci", ignore)]`?

Comment (FL33TW00D) [reply crates/ratchet-models/src/moondream/model.rs:241]: @timgestson Agreed
